### PR TITLE
Tables

### PR DESCRIPTION
**Pull request by bertiebaggio**
This PR adds a new command, `!table`, to covbot. 

**Example**:

covbot-table-example

It is a work in progress, and trusts the user knows what she's doing.

**Limitations**: 
 - doesn't check for duplicate entries (easy to do if desired)
 - calculates potentially nonsensical totals (if someone requests subregions)
 - aborts on any multiple matches (cf `len(matches) > 1` in `!cases`; can be made to warn and continue instead)

It duplicates a bunch of code which I am happy to refactor if this PR is notionally accepted for merging.

### CODE REVIEW

**Comment by bertiebaggio** [comment]
This looks bad on mobile:

covidtablemobile
*(thanks to Rahul for pointing this out)*

The original plan was to render the tables in preformatted text anyway, there are several python modules for this, see https://stackoverflow.com/a/26937531/4647265

**Comment by pwr22** [comment]
Looks great so far! :+1: on having a good fallback representation :thinking:

**Comment by bertiebaggio** [comment]
I settled on [tabulate](https://pypi.org/project/tabulate/) for the tables as it seems the most featureful and up to date. This does of course add another external dependency, but in my opinion that's better than rolling one's own.

**Example**:

covbottabulate

**Comment by pwr22** [comment]
External dependencies are no problem :+1: .

**Comment by bertiebaggio** [comment]
Grand, thanks. I'll continue to work on this to refactor, and will probably add a shorter table in too for the benefit of mobile as the text one is still only so-so at best.

covbotmobile2

*(cheers to JourneymanGeek for the screenshot)*

**Comment by pwr22** [comment]
Heh, yeah it's gonna be a pain to get a text based tabular format looking good on mobile :thinking:. Let me know when you're ready to get this merged :+1: . I'm sure the users will appreciate it :smile: .

**Comment by bertiebaggio** [comment]
### Small tables

Yup, mobile-friendly tables will take some thinking! It would be great if there was JS/CSS support, could do some column toggling/resizing or similar. But I digress. 

It looks like we have 70 characters before the table wraps, I will see if there's something that can be done to fit useful data in that space. My thought is a combination of: truncating/shortening locations, (maybe) leaving off % columns, and displaying numbers with k/M/B. Having a function that does just that may be of use in general for further down the line when numbers get that far...

### Merging

Thanks- my plan of attack today is:
 

- [x] merge in your changes from `master`
- [x] handle missing details (recoveries, deaths)
- [x] add the `!tablesmall` command (can rename this and the other if you prefer- ie `!smalltable` `!htmltable`)
- [x] test those with a small group of users 
- [x] refactor

**Comment by bertiebaggio** [comment]
<del>Some `!tablesmall` examples at 70 cols:</del>

**Edit(+2h)**: Never mind, clients aren't terribly consistent. My own test RiotX gets 40 monospaced chars on one line.

However, after much hair-pulling I have discovered that part of the issue is that mobile clients don't like to render markdown or HTML in `m.notice` events. 

Sending as `m.text` gives a horizontal scroll, which may be preferable to hard-limiting to 40 chars.

File: covbot.py
Comment (bertiebaggio) [inline covbot.py:8]: Not needed in the rest of the code as far as I can tell (nor pprint below)

File: covbot.py
Comment (bertiebaggio) [inline covbot.py:40]: Not sure how clear this is for end users?

There is also a case to be made for having multiple arguments so combinations are possible (eg `!table tiny html Loc1;Loc2;Loc3`), but I wanted to keep it as similar to `!cases` as possible

**Comment by bertiebaggio** [comment]
I think we're about at a good point to merge this feature branch, with your feedback / approval of course.

### Main additions

New commands (all take semicolon-delimited list of locations):

- `!table` - default table, outputting location/cases/sick/recovered/deaths
- `!tablesmall` - as above but limits location to 12 characters to reduce horizontal scroll
- `!tabletiny` - only outputs location and cases
- `!tablehtml` - outputs in HTML (not rendered in mobile clients at this time)

The above commands will elide columns where all data is missing (eg PHE cities).

### Minor

- removes unneeded imports (`requests` and `pprint`) (7bd3a223b883812dcb6591fe3a8a7c4cdf066beb)

### New functions

- `_short_location()`: returns shortened *location*, at max *length* characters
- `_get_multiple_locations()`: returns dict of dicts in form *{location1: {data1}, location2: {data2}}*
- `_locations_table()`: returns string with tablulated data for bot response
- `_respond_formatted()`:  bot output in `m.text` format with formatted data (cf: `_respond()`) -- needed for mobile client formatting

File: covbot.py
Comment (bertiebaggio) [inline covbot.py:394]: Ideally these would match the `!cases` 'not found' / 'multiple results' text.

That would be a simple case of reusing the strings, but I didn't wan to alter code outside of the remit of the table commands without prior approval.

**Comment by bertiebaggio** [comment]
### Output examples

#### Table

covtable1

#### Small table

covtable2

### Tiny Table

covtable3

## HTML Table

covtable4

### Mobile

covbottabletiny

File: covbot.py
Comment (pwr22) [reply covbot.py:394]: Np, I'll sort it out after the merge :+1: .

**Comment by pwr22** [comment]
Will look at merging this later today :).

**Comment by bertiebaggio** [comment]
Thanks! I am happy to make further changes if you'd like.

File: covbot.py
Comment (pwr22) [reply covbot.py:40]: Yeah, my preference would be to keep things simple

**Review by pwr22 — APPROVED** [review]
LGTM

**Comment by bertiebaggio** [comment]
Thanks for your time on this :)

**Pull request merged**
